src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -22,31 +22,22 @@ internal static partial class SpanHelpers
         /// </summary>
         public static unsafe void CopyTo<T>(ref T dst, int dstLength, ref T src, int srcLength)
         {
-            IntPtr srcMinusDst = Unsafe.ByteOffset<T>(ref dst, ref src);
-            bool srcGreaterThanDst = (sizeof(IntPtr) == sizeof(int)) ? srcMinusDst.ToInt32() >= 0 : srcMinusDst.ToInt64() >= 0;
-            IntPtr tailDiff;
+            Debug.Assert(dstLength != 0);
 
-            if (srcGreaterThanDst)
-            {
-                // If the start of source is greater than the start of destination, then we need to calculate
-                // the different between the end of destination relative to the start of source.
-                tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref dst, dstLength), ref src);
-            }
-            else
-            {
-                // If the start of source is less than the start of destination, then we need to calculate
-                // the different between the end of source relative to the start of destunation.
-                tailDiff = Unsafe.ByteOffset<T>(ref Unsafe.Add<T>(ref src, srcLength), ref dst);
-            }
+            IntPtr srcByteCount = Unsafe.ByteOffset(ref src, ref Unsafe.Add(ref src, srcLength));
+            IntPtr dstByteCount = Unsafe.ByteOffset(ref dst, ref Unsafe.Add(ref dst, dstLength));
+
+            IntPtr diff = Unsafe.ByteOffset(ref src, ref dst);
+
+            bool isOverlapped = (sizeof(IntPtr) == sizeof(int))
+                ? (uint)diff < (uint)srcByteCount || (uint)diff > ~(uint)dstByteCount + 1
```

Comment:
Can `~(uint)dstByteCount + 1` be `(uint)(-(int)dstByteCount)` instead? It should be one less instruction once JITed.